pyenv-win/bin/pyenv.bat

```diff
@@ -1,16 +1,93 @@
 @echo off
-setlocal enableDelayedExpansion
+setlocal
 chcp 1250 >nul
 
-IF EXIST "%~dp0"..\exec.bat (
-   del /F /Q "%~dp0"..\exec.bat >nul
+set "pyenv=cscript //nologo "%~dp0..\libexec\pyenv.vbs""
+
+rem if 'pyenv' called alone, then run pyenv.vbs
+if [%1]==[] (
+  %pyenv%
+  exit /b
 )
 
-call cscript //nologo "%~dp0"..\libexec\pyenv.vbs %*
-IF EXIST "%~dp0"..\exec.bat (
-   call "%~dp0"..\exec.bat
+rem use pyenv.vbs to aid resolving absolute path of "active" version into 'bindir'
+for /f %%i in ('%pyenv% version') do call :normalizepath "%~dp0..\versions\%%i" bindir
+
+rem all help implemented as plugin
+if /i [%2]==[--help] goto :plugin
+
+rem let pyenv.vbs handle these
+set commands=rehash global local version vname version-name versions commands shims which whence help --help
+for %%a in (%commands%) do (
+  if /i [%1]==[%%a] (
+    %pyenv% %*
+    exit /b
+  )
+)
+
+rem jump to plugin or fall to exec
+if /i not [%1]==[exec] goto :plugin
+
+
+rem ====================================================================================
+:exec
+
+shift
+
+for /f "tokens=1,2 delims=/" %%i in ("%1") do set "exepath=%%i" & set "exe=%%j"
+if [%exe%]==[] (set "exe=%exepath%") else (set "exe=%bindir%\%exepath%\%exe%")
+set "exe=cmd /c "%exe%""
+goto :run
+
+
+
+
+rem ====================================================================================
+:plugin
+
+set "exe=%~dp0..\libexec\pyenv-%1"
+call :normalizepath %exe% exe
+
+if exist "%exe%.bat" (
+  set "exe=endlocal & call "%exe%.bat""
```

Comment:
It seems this `endlocal` does not work properly.

Tests pass locally on my laptop because I use a `poetry` virtual env that updates the `PATH` so that when I run `pytest` it will execute the right `pytest.exe` directly.

In Travis CI, we use pyenv to run pytest. In the logs, it says that the arguments to pyenv start with `-m`.
This is because the call to `python -m pytest -v -s tests` actually uses `python.bat` in pyenv shims.
If calling `pyenv-shell.bat` in the unit tests still get `-m` as first argument, I guess this means `params` variable is not reset.

If I empty `params` variable before `:paramloop` (_but this is a dirty fix for me_), only one test fail: `test_shell_set_installed_version`. I guess this test fails for the same reason because `PYENV_VERSION` is not properly set because of `setlocal/endlocal` scopes.

Reply:
The test's expose an edge case when `pyenv.bat` indirectly calls back into itself; I would be surprised of this happens during normal usage.

Regardless, simply because of how environment variables are the only kind in '.bat' files, along with how setlocal/endlocal work, in the case of reentrancy, `params` and `bindir` need to be explicitly set to empty; so, not really a "dirty" fix, but the only fix.

I had previously stated that with this PR, `.bat` type plugins would have PATH prefixed with the "active" version path, but this was not true. In fact, it is not really possible to set PATH only for the duration of `exec` or the plugin, while still allowing the plugin to alter the calling shell's environment variables, which the `shell` command must be able to do.

I've special cased processing `.cmd` plugins, and the context for running things now is as follows:
* For the `exec` command, and for `.vbs`, and `.cmd` plugins, PATH is prefixed with the "active" python version, but none of them can alter any shell environment variables of the calling process/shell.
* For `.bat` plugins, PATH is **_not_** prefixed, but such plugins can modify the calling shell's environment variables, like pyenv-shell.bat does (as long they don't themselves call `setlocal`).